### PR TITLE
Wasmate: Support offsets on load/store operations

### PR DESCRIPTION
**Pull request by dschuff**

### CODE REVIEW

**Comment by dschuff** [comment]
@jfbastien @sunfishcode 
Here's one possible way to support mem offsets in .s syntax?

File: prototype-wasmate/test/store-trunc.s [hunk 1]
```diff
@@ -46,3 +46,11 @@ func_end4:
 	.size	trunc_i32_i64, func_end4-trunc_i32_i64
 
 
+	.globl	trunc_i8_i32_off
+	.type	trunc_i8_i32_off,@function
+trunc_i8_i32_off:
+	.param  	i32, i32
+	i32.store8	offset=1234, $discard=, $0, $1
```
Comment (sunfishcode) [inline prototype-wasmate/test/store-trunc.s]: This syntax will work, though it's not very assembly-like :-}.

Unfortunately, the more assembly-like things one might think of, perhaps something like the AT&T/MIPS/Power/etc.-like `1234($0)`, will require a little more tablegen hacking to implement (code roughly corresponding to X86MemOperand in X86, but with a little less hairiness) . It's something I think would be good to do, though if you'd like to defer it for now to focus more on the other parts that's fine with me.

Comment (jfbastien) [inline prototype-wasmate/test/store-trunc.s]: Agreed. `1234($0)` is nicer, but short-term we may want to stick with what @dschuff proposes. Maybe file a bug to fix later, if we go with the quick hack?

Comment (sunfishcode) [inline prototype-wasmate/test/store-trunc.s]: Wait, actually, it seems like we should be able to do this without significantly more tablegen hacking, just by using the AsmStrings. eg.

```
  "i32.load\t$dst, $offset($addr)"
```

seems like it ought to work. @dschuff, unless there are any complications, would you mind going with that?

**Comment by dschuff** [comment]
OK, I tried it out and it is pretty straight forward with the asm strings, so I've updated.

**Comment by jfbastien** [comment]
lgtm

File: prototype-wasmate/wasmate.py [hunk 2]
```diff
@@ -322,6 +322,16 @@ def handle_label(self, labelname):
                 out.indent()
 
     def handle_mnemonic(self, command, args):
+        # Handle address arguments of stores which have offsets.
+        # Make the offset part of the command instead of an arg, otherwise
+        # sexprify will interpret 'offset' as a label and prepend a '$'
+        if 'load' in command or 'store' in command:
+          addr_arg = 1 if 'load' in command else 1
```
Comment (sunfishcode) [inline prototype-wasmate/wasmate.py]: Is there a python subtlety here or is this just `addr_arg = 1`?

**Comment by sunfishcode** [comment]
lgtm

File: prototype-wasmate/wasmate.py (hunk 2 above)
Comment (dschuff) [inline prototype-wasmate/wasmate.py]: No, it's just a line I forgot to delete. addr_arg isn't even used.

**Comment by dschuff** [comment]
Interesting, if you force-push a branch and replace a commit with another one, all the comments on that commit disappear forever. I suppose I just shouldn't rewrite history like that.

**Pull request merged**